### PR TITLE
seperate process from main thread structure

### PR DESCRIPTION
**Pull request by paulcallen**

### CODE REVIEW

File: include/myst/thread.h [hunk 1]
```diff
@@ -456,6 +478,7 @@ int myst_set_thread_name(myst_thread_t* thread, const char* n);
 long myst_call_on_stack(void* stack, long (*func)(void* arg), void* arg);
 
 /* Send SIGHUP to child processes of given process thread */
-int myst_send_sighup_child_processes(myst_thread_t* process_thread);
+int myst_send_sighup_child_processes(
+    myst_process_t* MYST_PROCESS_MAX_STACK_SIZE);
```
Comment (vtikoo) [inline include/myst/thread.h:482]: Probably a typo?

Comment (paulcallen) [reply include/myst/thread.h]: fixed

File: include/myst/thread.h [hunk 2]
```diff
@@ -292,8 +314,13 @@ MYST_INLINE bool myst_valid_thread(const myst_thread_t* thread)
 }
 
 myst_thread_t* myst_thread_self(void);
+myst_process_t* myst_process_self(void);
 
-void myst_zombify_thread(myst_thread_t* thread);
+void myst_zombify_process(myst_process_t* process);
+MYST_INLINE bool myst_is_zombied_process(myst_process_t* process)
+{
+    return (process->main_process_thread == NULL) ? true : false;
```
Comment (vtikoo) [inline include/myst/thread.h:322]: For my understanding, the `myst_process_t` associated with the zombied process is free'd by the parent?

File: include/myst/thread.h [hunk 3]
```diff
@@ -96,18 +86,99 @@ struct myst_thread
...
+        /* process-wide protection for manipulating signals that use either mask
+         * or sigactions*/
+        myst_spinlock_t lock;
+
+        // the signal handles registered through sigaction and
+        // shared by threads in the prcoess.
+        posix_sigaction_t* sigactions;
+
+        /* The mask of blocked signals */
+        uint64_t mask;
```
Comment (vtikoo) [inline include/myst/thread.h:160]: Acc to the man page, it looks the mask is per-thread.
https://man7.org/linux/man-pages/man2/sigprocmask.2.html

Comment (paulcallen) [reply include/myst/thread.h]: yeah! I discovered that the hard way! Fixed

File: include/myst/thread.h (hunk 2 above)
Comment (paulcallen) [reply include/myst/thread.h:322]: process is only freed in myst_wait and during zombie shutdown

Comment (paulcallen) [reply include/myst/thread.h:322]: So what you say can be interpreted as only the parent, as only the parent can do a waitpid() on a child process.

**Review by mingweishih — APPROVED** [review]

**Pull request merged**